---
title: "ssw is on CRAN"
author: "Nan Xiao"
date: "2024-09-08T19:07:00"
slug: ssw-0-2-0
categories: []
tags:
  - R
  - R packages
  - bioinformatics
  - sequence alignment
  - Smith-Waterman algorithm
  - SIMD
bibliography: "ssw.bib"
meta_img: "image/nicholas-bartos-YwSxLoiXZFU-unsplash.jpg"
description: >
  ssw-r provides an R interface to the SSW C library and ssw-py for
  efficient sequence alignment using the Smith-Waterman algorithm.
---



<div class="float">
<img src="/image/nicholas-bartos-YwSxLoiXZFU-unsplash.jpg" alt="Cape Cod sand dunes on a cloudy day. Photo by Nicholas Bartos." />
<div class="figcaption">Cape Cod sand dunes on a cloudy day.
Photo by <a href="https://unsplash.com/photos/YwSxLoiXZFU">Nicholas Bartos</a>.</div>
</div>
<p>I’m excited to share that my R package <a href="https://nanx.me/ssw-r/">ssw</a>
is now available on CRAN. This package began as a weekend project in 2020.
ssw offers an R interface for
<a href="https://github.com/mengyao/Complete-Striped-Smith-Waterman-Library">SSW</a>
<span class="citation">(<a href="#ref-zhao2013ssw">Zhao et al. 2013</a>)</span>, a high-performance C/C++ implementation of the Smith-Waterman
algorithm for sequence alignment using SIMD.</p>
<p>You can install ssw from CRAN with:</p>
<pre class="r"><code>install.packages(&quot;ssw&quot;)</code></pre>
<p>For clarity, I will refer the R package as <strong>ssw-r</strong> from now on.
ssw-r currently wraps <a href="https://pypi.org/project/ssw-py/">ssw-py</a> via reticulate.
Assuming you have a recent version of Python installed,
you can set up ssw-py using the following helper function:</p>
<pre class="r"><code>ssw::install_ssw_py()</code></pre>
<p>This will install ssw-py into a virtual environment named <code>r-ssw-py</code>
by default, making it easily discoverable and importable by the R package.</p>
<div id="why-ssw" class="section level2">
<h2>Why SSW?</h2>
<p>I first learned the Smith-Waterman algorithm during a biomathematics
summer school course taught by Professor Michael Waterman in 2014.
Later, I realized how fundamental this algorithm is across bioinformatics.
However, its computational intensity can be a bottleneck in large-scale
genomic analyses. This is where the SSW C library comes into play.</p>
<p>The SSW C library accelerates Smith-Waterman alignment by leveraging
SIMD (Single Instruction, Multiple Data) instructions,
which allow for efficient vectorization on modern CPUs.
Besides the speedup, SSW provides detailed alignment information,
including suboptimal alignment scores, which makes it useful for tasks
ranging from short-read alignment to protein database search.
Importantly, it is a C library—one can easily integrate it into their
bioinformatics workflows without introducing unnecessary dependencies.</p>
<p>With ssw-r, we bridge the gap between high-performance SSW C library
and the rich ecosystem of R’s bioinformatics tools (for example,
Bioconductor). This potentially enables faster sequence analyses
in many tasks.</p>
</div>
<div id="examples" class="section level2">
<h2>Examples</h2>
<p>Let’s look at the two main functions of ssw-r: <code>align()</code> and <code>force_align()</code>.
For more examples, check out the <a href="https://nanx.me/ssw-r/articles/ssw.html">vignette</a>.</p>
<pre class="r"><code>library(&quot;ssw&quot;)</code></pre>
<p>First, align a short query sequence to a longer reference:</p>
<pre class="r"><code>(x &lt;- &quot;GATTACA&quot; |&gt; align(reference = &quot;CGGCTCTTGATTACAGGGTCT&quot;))</code></pre>
<pre class="text"><code>CIGAR start index 8: 7M
optimal_score: 14
sub-optimal_score: 0
target_begin: 8	target_end: 14
query_begin: 0
query_end: 6

Target:        8    GATTACA    14
                    |||||||
Query:         0    GATTACA    6</code></pre>
<pre class="r"><code>x$alignment$optimal_score
#&gt; [1] 14
x$alignment$sub_optimal_score
#&gt; [1] 0</code></pre>
<p>This example aligns a 7-base query to a 21-base reference.
The match is identified with the alignment details,
including the CIGAR string and the optimal alignment score.</p>
<p>In cases where we want to enforce strict alignment conditions,
such as avoiding gaps, we can use the <code>force_align()</code> function.</p>
<pre class="r"><code>(y &lt;- &quot;ACTG&quot; |&gt; force_align(reference = &quot;TTTTCTGCCCCCACG&quot;))</code></pre>
<pre class="text"><code>CIGAR start index 4: 3M
optimal_score: 6
sub-optimal_score: 0
target_begin: 4	target_end: 6
query_begin: 1
query_end: 3

Target:        4    CTG    6
                    |||
Query:         1    CTG    3</code></pre>
<p>To view the full alignment results without truncation, use <code>formatter()</code>:</p>
<pre class="r"><code>(y |&gt; formatter())
#&gt; [[1]]
#&gt; [1] &quot;TTTTCTGCCCCCACG&quot;
#&gt;
#&gt; [[2]]
#&gt; [1] &quot;   ACTG&quot;</code></pre>
<p>You can also pretty-print the formatted results directly:</p>
<pre class="r"><code>y |&gt; formatter(print = TRUE)
#&gt; TTTTCTGCCCCCACG
#&gt;    ACTG</code></pre>
</div>
<div id="acknowledgments" class="section level2">
<h2>Acknowledgments</h2>
<p>I would like to thank Mengyao Zhao for creating the original SSW library
and Nick Conway for maintaining the Python interface, ssw-py.
If you use ssw-r in your work, please consider citing these foundational projects.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<style type="text/css">
.references { font-size: 1.125rem; }
</style>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-zhao2013ssw" class="csl-entry">
Zhao, Mengyao, Wan-Ping Lee, Erik P Garrison, and Gabor T Marth. 2013. <span>“SSW Library: An <span>SIMD</span> Smith-Waterman <span>C/C++</span> Library for Use in Genomic Applications.”</span> <em><span>PloS ONE</span></em> 8 (12): e82138. <a href="https://doi.org/10.1371/journal.pone.0082138">https://doi.org/10.1371/journal.pone.0082138</a>.
</div>
</div>
</div>
