---
title: "Software"
menu: "main"
---

*  *  *  *

## R packages

I build R packages to improve my workflow.

### Infrastructure and automation

- [pkglite](https://merck.github.io/pkglite/)\
  Compact package representations\
  [CRAN](https://cran.r-project.org/package=pkglite) |
  [GitHub](https://github.com/Merck/pkglite) |
  [Paper](https://doi.org/10.1177/17407745221123244)

### Statistical machine learning

- [msaenet](https://nanx.me/msaenet/)\
  Multi-step adaptive estimation for sparse regressions\
  [CRAN](https://cran.r-project.org/package=msaenet) |
  [GitHub](https://github.com/nanxstats/msaenet) |
  [Paper](https://nanx.me/papers/msaenet.pdf)
- [stackgbm](https://nanx.me/stackgbm/)\
  Minimalist implementation of model stacking for gradient boosted tree models
  built by xgboost, lightgbm, and catboost\
  [GitHub](https://github.com/nanxstats/stackgbm)
- [oneclust](https://nanx.me/oneclust/)\
  Maximum homogeneity clustering for univariate data\
  [CRAN](https://cran.r-project.org/package=oneclust) |
  [GitHub](https://github.com/nanxstats/oneclust)
- [logreg](https://nanx.me/logreg/)\
  Regularized logistic regressions with computational graphs\
  [GitHub](https://github.com/nanxstats/logreg)
- [OHPL](https://OHPL.io/doc/)\
  Ordered homogeneity pursuit lasso for group variable selection\
  [CRAN](https://cran.r-project.org/package=OHPL) |
  [GitHub](https://github.com/nanxstats/OHPL) |
  [Paper](https://nanx.me/papers/OHPL.pdf)
- [RECA](https://nanx.me/RECA/)\
- Relevant component analysis for supervised distance metric learning\
  [CRAN](https://cran.r-project.org/package=RECA) |
  [GitHub](https://github.com/nanxstats/RECA)
- [enpls](https://nanx.me/enpls/)\
  Ensemble partial least squares regression\
  [CRAN](https://cran.r-project.org/package=enpls) |
  [GitHub](https://github.com/nanxstats/enpls)

### Statistical graphics and computing

- [ggsci](https://nanx.me/ggsci/)\
  Scientific journal and sci-fi themed color palettes for ggplot2\
  [CRAN](https://cran.r-project.org/package=ggsci) |
  [GitHub](https://github.com/nanxstats/ggsci)
- [liftr](https://liftr.me/)\
  Containerize R Markdown documents\
  [CRAN](https://cran.r-project.org/package=liftr) |
  [GitHub](https://github.com/nanxstats/liftr) |
  [Paper](https://doi.org/10.32614/RJ-2020-007)
- [hdnom](https://nanx.me/hdnom/)\
  Benchmarking and visualization toolkit for penalized Cox models\
  [CRAN](https://cran.r-project.org/package=hdnom) |
  [GitHub](https://github.com/nanxstats/hdnom) |
  [Paper](https://doi.org/10.1101/065524)

### Bioinformatics and cheminformatics

- [protr](https://nanx.me/protr/)\
  Protein sequence feature extraction\
  [CRAN](https://cran.r-project.org/package=protr) |
  [GitHub](https://github.com/nanxstats/protr) |
  [Paper](https://nanx.me/papers/protr.pdf)
- [Rcpi](https://nanx.me/Rcpi/)\
  Feature extraction for proteins, compounds, and their interactions\
  [Bioconductor](https://bioconductor.org/packages/Rcpi) |
  [GitHub](https://github.com/nanxstats/Rcpi) |
  [Paper](https://nanx.me/papers/Rcpi.pdf)
- [tidychem](https://nanx.me/tidychem/)\
  Read, preprocess, and featurize chemical data using RDKit\
  [GitHub](https://github.com/nanxstats/tidychem)
- [grex](https://nanx.me/grex/)\
  Gene ID mapping for Genotype-Tissue Expression (GTEx) data\
  [CRAN](https://cran.r-project.org/package=grex) |
  [GitHub](https://github.com/nanxstats/grex)
- [cats](https://nanx.me/cats/)\
  Joint power analysis for non-symmetric two-stage case-control designs\
  [GitHub](https://github.com/nanxstats/cats)
- [ssw-r](https://nanx.me/ssw-r/)\
  R interface for SSW, a fast implementation of the Smith-Waterman algorithm using SIMD\
  [GitHub](https://github.com/nanxstats/ssw-r)
- [tidycwl](https://sbg.github.io/tidycwl/)\
  Tidy Common Workflow Language tools and workflows\
  [CRAN](https://cran.r-project.org/package=tidycwl) |
  [GitHub](https://github.com/sbg/tidycwl) |
  [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [biocompute](https://sbg.github.io/biocompute/)\
  Create and manipulate BioCompute Objects\
  [CRAN](https://cran.r-project.org/package=biocompute) |
  [GitHub](https://github.com/sbg/biocompute) |
  [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [sevenbridges-r](https://sbg.github.io/sevenbridges-r/)\
  Seven Bridges API client\
  [Bioconductor](https://bioconductor.org/packages/sevenbridges) |
  [GitHub](https://github.com/sbg/sevenbridges-r)
- [sbpack-r](https://nanx.me/sbpack-r/)\
  Pack and upload CWL apps to Seven Bridges powered platforms\
  [GitHub](https://github.com/nanxstats/sbpack-r)

## Web applications

### JavaScript apps

- [Deep Learning GPU Selector](https://nanx.me/gpu/)\
  Discover the best GPU for your deep learning needs by answering a few yes/no questions.\
  [GitHub](https://github.com/nanxstats/gpu) |
  [Blog post](https://nanx.me/blog/post/gpu-selector/)

### Shiny apps

Shiny apps, widgets, and templates for interactive data analysis.

- [shiny-loading-skeleton](https://github.com/nanxstats/shiny-loading-skeleton)\
  Shiny app template with an animated, fully customizable skeleton loader.\
  [Demo](https://nanx.shinyapps.io/shiny-loading-skeleton/)
- [shiny-fcp-loader](https://github.com/nanxstats/shiny-fcp-loader)\
  Shiny app template with a loading screen that aims to minimize the time to first contentful paint (FCP).\
  [Demo](https://nanx.shinyapps.io/shiny-fcp-loader/)
- [shiny-highlight-rmarkdown](https://github.com/nanxstats/shiny-highlight-rmarkdown)\
  Syntax highlighting for R Markdown output in Shiny based on highlight.js\
  [Demo](https://nanx.shinyapps.io/shiny-highlight-rmarkdown/)
- [exp2vec-shiny](https://github.com/nanxstats/exp2vec-shiny)\
  Shiny app for gene neighbors and gene analogies discovery with tissue-specific gene embeddings training on GTEx data.\
  [Demo](https://nanx.shinyapps.io/exp2vec/)
- [bco-app](https://github.com/sbg/bco-app)\
  Create, validate, and browse BioCompute Objects. Advanced track winner of the [precisionFDA challenge on BioCompute Objects](https://precision.fda.gov/challenges/7/results).\
  [Documentation](https://sbg.github.io/bco-app/) |
  [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [protrweb](https://github.com/nanxstats/protrweb)\
  Compute protein sequence-derived descriptors\
  [Mirror 1](https://nanx.app/protr/) |
  [Mirror 2](https://nanx.shinyapps.io/protrweb/) |
  [Paper](https://nanx.me/papers/protr.pdf)
- [targetnet](https://github.com/nanxstats/targetnet)\
  Drug target identification with large-scale binding affinities data\
  [Mirror 1](https://nanx.app/targetnet/) |
  [Mirror 2](https://nanx.shinyapps.io/targetnet/)
- [imgsvd](https://github.com/nanxstats/imgsvd)\
  Image compression via singular value decomposition\
  [Mirror 1](https://nanx.app/imgsvd/) |
  [Mirror 2](https://nanx.shinyapps.io/imgsvd/)
- [hdnom-app](https://github.com/nanxstats/hdnom-app)\
  Benchmarking and visualization of penalized Cox models\
  [Mirror 1](https://nanx.app/hdnom/) |
  [Mirror 2](https://nanx.shinyapps.io/hdnom-app/)
- [conjugate-normal-umkv](https://github.com/nanxstats/conjugate-normal-umkv)\
  A demo of [Bayesian inference for normal mean (known variance)](https://stephens999.github.io/fiveMinuteStats/shiny_normal_example.html)\
  [Mirror 1](https://nanx.app/conjugate-normal-umkv/) |
  [Mirror 2](https://nanx.shinyapps.io/conjugate-normal-umkv/)
- [signify](https://github.com/nanxstats/signify)\
  Make your (>0.05) *p*-values "sound significant"\
  [Mirror 1](https://nanx.app/signify/) |
  [Mirror 2](https://nanx.shinyapps.io/signify/)
- [cwl-formatter](https://github.com/nanxstats/cwl-formatter)\
  Format Common Workflow Language (CWL) apps\
  [Mirror 1](https://nanx.app/cwl-formatter/)
- [cwl-exploder](https://github.com/nanxstats/cwl-exploder)\
  Unpack monolithic CWL workflows\
  [Mirror 1](https://nanx.app/cwl-exploder/)
- [cwl-upgrader](https://github.com/nanxstats/cwl-upgrader)\
  Upgrade CWL apps from sbg:draft2 to v1.0\
  [Mirror 1](https://nanx.app/cwl-upgrader/)

## Data analysis pipelines

Reusable pipelines for analyzing real-world datasets in specific domains.

- [exp2vec](https://github.com/nanxstats/exp2vec)\
  Pipeline for training tissue-specific gene embeddings on GTEx data.
- [bcpm-msaenet](https://github.com/nanxstats/bcpm-msaenet)\
  Pipeline for the [precisionFDA challenge on brain cancer predictive modeling and biomarker discovery](https://precision.fda.gov/challenges/8/results).
  The 2nd place solution by predictive performance.
  Features multi-step adaptive SCAD-net and relaxed stability selection
  for feature selection, and a two-layer stacking tree model with xgboost,
  lightgbm, and catboost.
- [rankv](https://github.com/nanxstats/rankv)\
  Pipeline for the [precisionFDA challenge on detecting adverse event anomalies using FDA open data](https://precision.fda.gov/challenges/9).
  Features a rank aggregated signal detection method for
  Vaccine Adverse Event Reporting System (VAERS) data,
  with base signal detection methods including Gamma Poisson Shrinker,
  PRR, ROR, and BCPNN. [Website](https://nanx.me/rankv/).

## Code contributor

- [gMCPLite](https://github.com/Merck/gMCPLite)\
  Lightweight fork of gMCP for graph-based multiple comparison procedures\
  [CRAN](https://cran.r-project.org/package=gMCPLite)
- [gsDesign Shiny app](https://rinpharma.shinyapps.io/gsdesign/)\
  Group sequential clinical trial design\
  [Mirror 1](https://gsdesign.shinyapps.io/prod/) |
  [Mirror 2](https://gsdesign.shinyapps.io/devel/)
- [PDXNet Portal](https://portal.pdxnetwork.org/)\
  Patient-derived xenograft (PDX) model, data, workflow, and tool discovery\
  [Paper](https://doi.org/10.1093/narcan/zcac014)
- [ashr](https://github.com/stephens999/ashr)\
  Methods for adaptive shrinkage, using Empirical Bayes\
  [CRAN](https://cran.r-project.org/package=ashr) |
  [Paper](https://doi.org/10.1093/biostatistics/kxw041)
- [dml](https://github.com/terrytangyuan/dml)\
  Distance metric learning in R\
  [CRAN](https://cran.r-project.org/package=dml) |
  [Paper](https://doi.org/10.21105/joss.01036)

## Code recipes & awesome lists

- [awesome-shiny-extensions](https://github.com/nanxstats/awesome-shiny-extensions)\
  A curated list of R packages that offer extended UI or server components for Shiny.
- [deep-learning-recipes](https://github.com/nanxstats/deep-learning-recipes)\
  R implementation for selected machine learning methods with deep learning frameworks.

## Visual Studio Code extensions

- [vscode-textmate-rstheme](https://marketplace.visualstudio.com/items?itemName=nanxstats.textmate-rstheme)\
  A light theme for Visual Studio Code inspired by the TextMate (default) theme in RStudio IDE.\
  [GitHub repository](https://github.com/nanxstats/vscode-textmate-rstheme)

## Hugo/blogdown themes

- [hugo-tanka](https://github.com/nanxstats/hugo-tanka)\
  A minimalist theme for Hugo and blogdown\
  [Live Preview](https://nanx.me/hugo-tanka/) |
  [Hugo Themes](https://themes.gohugo.io/hugo-tanka/) |
  [blogdown book](https://bookdown.org/yihui/blogdown/other-themes.html)
- [hugo-renga](https://github.com/nanxstats/hugo-renga)\
  A clean Bootstrap 4 theme for Hugo and blogdown\
  [Live Preview](https://nanx.me/hugo-renga/)
- [hugo-nanx2020](https://github.com/nanxstats/hugo-nanx2020)\
  An opinionated Bootstrap 4 theme for Hugo and blogdown
- [hugo-nanx2021](https://github.com/nanxstats/hugo-nanx2021)\
  A minimalist Bootstrap 5 theme for Hugo and blogdown
- [hugo-nanx2022](https://github.com/nanxstats/hugo-nanx2022)\
  A minimalist Bootstrap 5 theme for Hugo and blogdown

## Collections

- [dockflow.org](https://dockflow.org)\
  Bioconductor workflow containerization with liftr\
  [GitHub](https://github.com/nanxstats/dockflow)
- [nanx.app](https://nanx.app)\
  A collection of shiny apps built by me\
  [GitHub](https://github.com/nanxstats/nanx.app)

## Archives

- [使用 R 和 Apache 开发 Web 应用程序](https://nanx.me/rapache/) (2011)
- [来自 Google 的 R 语言编码风格指南](https://nanx.me/rstyle/) (2011)
