---
title: "Software"
menu: "main"
---

*  *  *  *

<div class="software-toc">
  <div class="row g-0">
    <div class="col-md-6">
      <a href="#python-packages" class="btn btn-outline-nav w-100">
        <i class="fab fa-python"></i>
        <span>Python Packages</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#r-packages" class="btn btn-outline-nav w-100">
        <i class="fab fa-r-project"></i>
        <span>R Packages</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#web-applications" class="btn btn-outline-nav w-100">
        <i class="fas fa-globe"></i>
        <span>Web Applications</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#machine-learning-workflows" class="btn btn-outline-nav w-100">
        <i class="fas fa-share-nodes"></i>
        <span>Machine Learning Workflows</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#open-source-contributions" class="btn btn-outline-nav w-100">
        <i class="fas fa-code-merge"></i>
        <span>Open Source Contributions</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#awesome-lists-and-code-recipes" class="btn btn-outline-nav w-100">
        <i class="fas fa-list-ul"></i>
        <span>Awesome Lists</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#visual-studio-code-extensions" class="btn btn-outline-nav w-100">
        <i class="fas fa-puzzle-piece"></i>
        <span>VS Code Extensions</span>
      </a>
    </div>
    <div class="col-md-6">
      <a href="#hugoblogdown-themes" class="btn btn-outline-nav w-100">
        <i class="fas fa-wand-magic-sparkles"></i>
        <span>Hugo Themes</span>
      </a>
    </div>
  </div>
</div>

<div class="mt-5"></div>

## Python packages

- [tinytopics](https://nanx.me/tinytopics/)\
  GPU-accelerated topic modeling via sum-to-one constrained
  neural Poisson NMF\
  [PyPI](https://pypi.org/project/tinytopics/) |
  [GitHub](https://github.com/nanxstats/tinytopics)
- [tinyvdiff](https://nanx.me/tinyvdiff/)\
  Minimalist visual regression testing plugin for pytest\
  [PyPI](https://pypi.org/project/tinyvdiff/) |
  [GitHub](https://github.com/nanxstats/tinyvdiff)
- [py-pkglite](https://pharmaverse.github.io/py-pkglite/)\
  A simple framework for packing source packages written in any programming
  language into a text file and restoring them into the original directory
  structure\
  [PyPI](https://pypi.org/project/pkglite/) |
  [GitHub](https://github.com/pharmaverse/py-pkglite)
- [rtflite](https://pharmaverse.github.io/rtflite/)\
  Lightweight RTF composer for Python\
  [PyPI](https://pypi.org/project/rtflite/) |
  [GitHub](https://github.com/pharmaverse/rtflite)

## R packages

### Infrastructure and automation

- [pkglite](https://merck.github.io/pkglite/)\
  Compact package representations\
  [CRAN](https://cran.r-project.org/package=pkglite) |
  [GitHub](https://github.com/Merck/pkglite) |
  [Paper](https://doi.org/10.1177/17407745221123244)
- [pkgdown.offline](https://nanx.me/pkgdown.offline/)\
  Build pkgdown websites without an internet connection\
  [CRAN](https://cran.r-project.org/package=pkgdown.offline) |
  [GitHub](https://github.com/nanxstats/pkgdown.offline)

### Statistical machine learning

- [msaenet](https://nanx.me/msaenet/)\
  Multi-step adaptive estimation for sparse regressions\
  [CRAN](https://cran.r-project.org/package=msaenet) |
  [GitHub](https://github.com/nanxstats/msaenet) |
  [Paper](https://nanx.me/papers/msaenet.pdf)
- [stackgbm](https://nanx.me/stackgbm/)\
  Minimalist implementation of model stacking for gradient boosted tree models
  built by xgboost, lightgbm, and catboost\
  [CRAN](https://cran.r-project.org/package=stackgbm) |
  [GitHub](https://github.com/nanxstats/stackgbm)
- [oneclust](https://nanx.me/oneclust/)\
  Maximum homogeneity clustering for univariate data\
  [CRAN](https://cran.r-project.org/package=oneclust) |
  [GitHub](https://github.com/nanxstats/oneclust)
- [logreg](https://nanx.me/logreg/)\
  Regularized logistic regressions with computational graphs\
  [GitHub](https://github.com/nanxstats/logreg)
- [OHPL](https://OHPL.io/doc/)\
  Ordered homogeneity pursuit lasso for group variable selection\
  [CRAN](https://cran.r-project.org/package=OHPL) |
  [GitHub](https://github.com/nanxstats/OHPL) |
  [Paper](https://nanx.me/papers/OHPL.pdf)
- [RECA](https://nanx.me/RECA/)\
  Relevant component analysis for supervised distance metric learning\
  [CRAN](https://cran.r-project.org/package=RECA) |
  [GitHub](https://github.com/nanxstats/RECA)
- [enpls](https://nanx.me/enpls/)\
  Ensemble partial least squares regression\
  [CRAN](https://cran.r-project.org/package=enpls) |
  [GitHub](https://github.com/nanxstats/enpls)

### Statistical graphics and computing

- [ggsci](https://nanx.me/ggsci/)\
  Scientific journal and sci-fi themed color palettes for ggplot2\
  [CRAN](https://cran.r-project.org/package=ggsci) |
  [GitHub](https://github.com/nanxstats/ggsci)
- [liftr](https://liftr.me/)\
  Containerize R Markdown documents\
  [CRAN](https://cran.r-project.org/package=liftr) |
  [GitHub](https://github.com/nanxstats/liftr) |
  [Paper](https://doi.org/10.32614/RJ-2020-007)
- [hdnom](https://nanx.me/hdnom/)\
  Benchmarking and visualization toolkit for penalized Cox models\
  [CRAN](https://cran.r-project.org/package=hdnom) |
  [GitHub](https://github.com/nanxstats/hdnom) |
  [Paper](https://doi.org/10.1101/065524)

### Bioinformatics and cheminformatics

- [protr](https://nanx.me/protr/)\
  Protein sequence feature extraction\
  [CRAN](https://cran.r-project.org/package=protr) |
  [GitHub](https://github.com/nanxstats/protr) |
  [Paper](https://nanx.me/papers/protr.pdf)
- [Rcpi](https://nanx.me/Rcpi/)\
  Feature extraction for proteins, compounds, and their interactions\
  [Bioconductor](https://bioconductor.org/packages/Rcpi) |
  [GitHub](https://github.com/nanxstats/Rcpi) |
  [Paper](https://nanx.me/papers/Rcpi.pdf)
- [ssw-r](https://nanx.me/ssw-r/)\
  R interface for SSW, a fast implementation of the Smith-Waterman algorithm using SIMD\
  [CRAN](https://cran.r-project.org/package=ssw) |
  [GitHub](https://github.com/nanxstats/ssw-r)
- [grex](https://nanx.me/grex/)\
  Gene ID mapping for Genotype-Tissue Expression (GTEx) data\
  [CRAN](https://cran.r-project.org/package=grex) |
  [GitHub](https://github.com/nanxstats/grex)
- [tidychem](https://nanx.me/tidychem/)\
  Read, preprocess, and featurize chemical data using RDKit\
  [GitHub](https://github.com/nanxstats/tidychem)
- [cats](https://nanx.me/cats/)\
  Joint power analysis for non-symmetric two-stage case-control designs\
  [GitHub](https://github.com/nanxstats/cats)
- [tidycwl](https://sbg.github.io/tidycwl/)\
  Tidy Common Workflow Language tools and workflows\
  [CRAN](https://cran.r-project.org/package=tidycwl) |
  [GitHub](https://github.com/sbg/tidycwl) |
  [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [biocompute](https://sbg.github.io/biocompute/)\
  Create and manipulate BioCompute Objects\
  [CRAN](https://cran.r-project.org/package=biocompute) |
  [GitHub](https://github.com/sbg/biocompute) |
  [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [sevenbridges-r](https://sbg.github.io/sevenbridges-r/)\
  Seven Bridges API client\
  [Bioconductor](https://bioconductor.org/packages/sevenbridges) |
  [GitHub](https://github.com/sbg/sevenbridges-r)
- [sbpack-r](https://nanx.me/sbpack-r/)\
  Pack and upload CWL apps to Seven Bridges powered platforms\
  [GitHub](https://github.com/nanxstats/sbpack-r)

## Web applications

### JavaScript apps

- [Deep Learning GPU Selector](https://nanx.me/gpu/)\
  Discover the best GPU for your deep learning needs by answering a few yes/no questions\
  [GitHub](https://github.com/nanxstats/gpu) |
  [Blog post](https://nanx.me/blog/post/gpu-selector/)
- [Word Cloud Generator (Batteries Included)](https://nanx.me/wordcloud/)\
  A fork version of Jason Davies' word cloud generator with color palette selection and updated default values\
  [GitHub](https://github.com/nanxstats/wordcloud) |
  [Blog post](https://nanx.me/blog/post/rethinking-word-cloud-generator/)
- [font-overlay](https://nanx.me/font-overlay/)\
  Interactively compare typesetting characteristics between fonts
  across customizable text, font size, letter spacing, and optical size.\
  [GitHub](https://github.com/nanxstats/font-overlay)

### Shiny apps

Shiny apps, widgets, and templates for interactive data analysis.

- [shiny-loading-skeleton](https://github.com/nanxstats/shiny-loading-skeleton)\
  Shiny app template with an animated, fully customizable skeleton loader\
  [Demo](https://nanx.shinyapps.io/shiny-loading-skeleton/)
- [shiny-fcp-loader](https://github.com/nanxstats/shiny-fcp-loader)\
  Shiny app template with a loading screen that aims to minimize the time to first contentful paint (FCP)\
  [Demo](https://nanx.shinyapps.io/shiny-fcp-loader/)
- [shiny-highlight-rmarkdown](https://github.com/nanxstats/shiny-highlight-rmarkdown)\
  Syntax highlighting for R Markdown output in Shiny based on highlight.js\
  [Demo](https://nanx.shinyapps.io/shiny-highlight-rmarkdown/)
- [exp2vec-shiny](https://github.com/nanxstats/exp2vec-shiny)\
  Shiny app for gene neighbors and gene analogies discovery with tissue-specific gene embeddings training on GTEx data\
  [Demo](https://nanx.shinyapps.io/exp2vec/)
- [bco-app](https://github.com/sbg/bco-app)\
  Create, validate, and browse BioCompute Objects\
  Advanced track winner of the [precisionFDA challenge on BioCompute Objects](https://precision.fda.gov/challenges/7/results)\
  [Documentation](https://sbg.github.io/bco-app/) |
  [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [protrweb](https://github.com/nanxstats/protrweb)\
  Compute protein sequence-derived descriptors\
  [Mirror 1](https://nanx.app/protr/) |
  [Mirror 2](https://nanx.shinyapps.io/protrweb/) |
  [Paper](https://nanx.me/papers/protr.pdf)
- [targetnet](https://github.com/nanxstats/targetnet)\
  Drug target identification with large-scale binding affinities data\
  [Mirror 1](https://nanx.app/targetnet/) |
  [Mirror 2](https://nanx.shinyapps.io/targetnet/)
- [imgsvd](https://github.com/nanxstats/imgsvd)\
  Image compression via singular value decomposition\
  [Mirror 1](https://nanx.app/imgsvd/) |
  [Mirror 2](https://nanx.shinyapps.io/imgsvd/)
- [hdnom-app](https://github.com/nanxstats/hdnom-app)\
  Benchmarking and visualization of penalized Cox models\
  [Mirror 1](https://nanx.app/hdnom/)
- [conjugate-normal-umkv](https://github.com/nanxstats/conjugate-normal-umkv)\
  A demo of [Bayesian inference for normal mean (known variance)](https://stephens999.github.io/fiveMinuteStats/shiny_normal_example.html)\
  [Mirror 1](https://nanx.app/conjugate-normal-umkv/) |
  [Mirror 2](https://nanx.shinyapps.io/conjugate-normal-umkv/)
- [signify](https://github.com/nanxstats/signify)\
  Make your (>0.05) *p*-values "sound significant"\
  [Mirror 1](https://nanx.app/signify/) |
  [Mirror 2](https://nanx.shinyapps.io/signify/)
- [cwl-formatter](https://github.com/nanxstats/cwl-formatter)\
  Format Common Workflow Language (CWL) apps\
  [Mirror 1](https://nanx.app/cwl-formatter/)
- [cwl-exploder](https://github.com/nanxstats/cwl-exploder)\
  Unpack monolithic CWL workflows\
  [Mirror 1](https://nanx.app/cwl-exploder/)
- [cwl-upgrader](https://github.com/nanxstats/cwl-upgrader)\
  Upgrade CWL apps from sbg:draft2 to v1.0\
  [Mirror 1](https://nanx.app/cwl-upgrader/)

## Machine learning workflows

- [exp2vec](https://github.com/nanxstats/exp2vec)\
  Workflow for training tissue-specific gene embeddings on GTEx data
- [bcpm-msaenet](https://github.com/nanxstats/bcpm-msaenet)\
  Workflow for the [precisionFDA challenge on brain cancer predictive modeling and biomarker discovery](https://precision.fda.gov/challenges/8/results).
  The 2nd place solution by predictive performance.
  Features multi-step adaptive SCAD-net and relaxed stability selection
  for feature selection, and a two-layer stacking tree model with xgboost,
  lightgbm, and catboost.
- [rankv](https://github.com/nanxstats/rankv)\
  Workflow for the [precisionFDA challenge on detecting adverse event anomalies using FDA open data](https://precision.fda.gov/challenges/9).
  Features a rank aggregated signal detection method for
  Vaccine Adverse Event Reporting System (VAERS) data,
  with base signal detection methods including Gamma Poisson Shrinker,
  PRR, ROR, and BCPNN.\
  [Website](https://nanx.me/rankv/)
- [ChromaClust](https://github.com/nanxstats/ChromaClust)\
  Image color topic modeling using
  [fastTopics](https://github.com/stephenslab/fastTopics).\
  The original idea was from my final project
  ([poster](https://nanx.me/posters/chromaclust-poster-hg48600.pdf))
  for the course HGEN 48600/STAT 35450.

## Open source contributions

- [gsDesign](https://keaven.github.io/gsDesign/)\
  Group sequential clinical trial design,
  largely as presented in Jennison and Turnbull (2000)\
  [CRAN](https://cran.r-project.org/package=gsDesign) |
  [GitHub](https://github.com/keaven/gsDesign) |
  [gsDesign: 15 Years of Development in 5 Minutes (4K 60fps)](https://www.youtube.com/watch?v=_ZUddQAhPFw)
- [gsDesign Shiny app](https://rinpharma.shinyapps.io/gsdesign/)\
  Web application for group sequential clinical trial design\
  [Mirror 1](https://gsdesign.shinyapps.io/prod/) |
  [Mirror 2](https://gsdesign.shinyapps.io/devel/)
- [gsDesign2](https://merck.github.io/gsDesign2/)\
  Group sequential design with non-constant effect\
  [CRAN](https://cran.r-project.org/package=gsDesign2) |
  [GitHub](https://github.com/Merck/gsDesign2)
- [simtrial](https://merck.github.io/simtrial/)\
  Clinical trial simulation\
  [CRAN](https://cran.r-project.org/package=simtrial) |
  [GitHub](https://github.com/Merck/simtrial)
- [gMCPLite](https://merck.github.io/gMCPLite/)\
  Lightweight fork of gMCP for graph-based multiple comparison procedures\
  [CRAN](https://cran.r-project.org/package=gMCPLite) |
  [GitHub](https://github.com/Merck/gMCPLite)
- [gMCPShiny](https://merck.github.io/gMCPShiny/)\
  Shiny app for graphical multiplicity control\
  [GitHub](https://github.com/Merck/gMCPShiny)
- [r2rtf](https://merck.github.io/r2rtf/)\
  Create production-ready Rich Text Format (RTF) tables and figures\
  [CRAN](https://cran.r-project.org/package=r2rtf) |
  [GitHub](https://github.com/Merck/r2rtf)
- [metalite](https://merck.github.io/metalite/)\
  Unified data structure for metadata information in clinical
  analysis & reporting (A&R), leveraging the
  Analysis Data Model (ADaM) datasets for consistent and
  accurate metadata representation\
  [CRAN](https://cran.r-project.org/package=metalite) |
  [GitHub](https://github.com/Merck/metalite)
- [metalite.ae](https://merck.github.io/metalite.ae/)\
  Analyzes adverse events in clinical trials using the metalite data structure.
  Simplifies the workflow to create production-ready tables, listings,
  and figures discussed in the adverse events analysis chapters of
  *R for Clinical Study Reports and Submission*.\
  [CRAN](https://cran.r-project.org/package=metalite.ae) |
  [GitHub](https://github.com/Merck/metalite.ae)
- [boxly](https://merck.github.io/boxly/)\
  Interactive box plot using plotly for clinical data analysis\
  [CRAN](https://cran.r-project.org/package=boxly) |
  [GitHub](https://github.com/Merck/boxly)
- [forestly](https://merck.github.io/forestly/)\
  Interactive forest plot for clinical trial safety analysis
  using metalite, reactable, plotly, and Analysis Data Model (ADaM) datasets\
  [CRAN](https://cran.r-project.org/package=forestly) |
  [GitHub](https://github.com/Merck/forestly)
- [PDXNet Portal](https://portal.pdxnetwork.org/)\
  Patient-derived xenograft (PDX) model, data, workflow, and tool discovery\
  [Paper](https://doi.org/10.1093/narcan/zcac014)
- [ashr](https://github.com/stephens999/ashr)\
  Methods for adaptive shrinkage, using Empirical Bayes\
  [CRAN](https://cran.r-project.org/package=ashr) |
  [Paper](https://doi.org/10.1093/biostatistics/kxw041)
- [dml](https://github.com/terrytangyuan/dml)\
  Distance metric learning in R\
  [CRAN](https://cran.r-project.org/package=dml) |
  [Paper](https://doi.org/10.21105/joss.01036)

## Awesome lists and code recipes

- [awesome-shiny-extensions](https://github.com/nanxstats/awesome-shiny-extensions)\
  A curated list of R packages that offer extended UI or server components for Shiny
- [awesome-webr](https://github.com/nanxstats/awesome-webr)\
  A curated list of awesome resources for learning WebR, a version of the
  statistical language R compiled for the browser and Node.js using WebAssembly
- [r-base-shortcuts](https://github.com/nanxstats/r-base-shortcuts)\
  A collection of lesser-known base R idioms and shortcuts for
  writing concise and fast base R code
- [r-rust-pkgs](https://github.com/nanxstats/r-rust-pkgs)\
  R packages using Rust on CRAN
- [r-future-recipes](https://github.com/nanxstats/r-future-recipes)\
  Guides and examples for the R future framework for parallel computing
- [r-windows-paths](https://github.com/nanxstats/r-windows-paths)\
  Key R toolchain paths for Windows systems
- [deep-learning-recipes](https://github.com/nanxstats/deep-learning-recipes)\
  R implementation for selected machine learning methods with deep learning frameworks
- [llm-cliches](https://github.com/nanxstats/llm-cliches)\
  A collection of commonly used clichés and phrases in Large Language Models (LLMs) outputs.

## Visual Studio Code extensions

- [vscode-textmate-rstheme](https://marketplace.visualstudio.com/items?itemName=nanxstats.textmate-rstheme)\
  A light theme for Visual Studio Code inspired by the TextMate (default) theme in RStudio IDE\
  [GitHub repository](https://github.com/nanxstats/vscode-textmate-rstheme)

## Hugo/blogdown themes

- [hugo-tanka](https://github.com/nanxstats/hugo-tanka)\
  A minimalist theme for Hugo and blogdown\
  [Live Preview](https://nanx.me/hugo-tanka/) |
  [Hugo Themes](https://themes.gohugo.io/hugo-tanka/) |
  [blogdown book](https://bookdown.org/yihui/blogdown/other-themes.html)
- [hugo-renga](https://github.com/nanxstats/hugo-renga)\
  A clean Bootstrap 4 theme for Hugo and blogdown\
  [Live Preview](https://nanx.me/hugo-renga/)
- [hugo-nanx2020](https://github.com/nanxstats/hugo-nanx2020)\
  An opinionated Bootstrap 4 theme for Hugo and blogdown
- [hugo-nanx2021](https://github.com/nanxstats/hugo-nanx2021)\
  A minimalist Bootstrap 5 theme for Hugo and blogdown
- [hugo-nanx2022](https://github.com/nanxstats/hugo-nanx2022)\
  A minimalist Bootstrap 5 theme for Hugo and blogdown
- [hugo-nanx2023](https://github.com/nanxstats/hugo-nanx2023)\
  A minimalist Bootstrap 5 theme for Hugo and blogdown

## Collections

- [dockflow.org](https://dockflow.org)\
  Bioconductor workflow containerization with liftr\
  [GitHub](https://github.com/nanxstats/dockflow)
- [nanx.app](https://nanx.app)\
  A collection of shiny apps built by me\
  [GitHub](https://github.com/nanxstats/nanx.app)

## Archives

- [使用 R 和 Apache 开发 Web 应用程序](https://nanx.me/rapache/) (2011)
- [来自 Google 的 R 语言编码风格指南](https://nanx.me/rstyle/) (2011)

<style>
.markdown ul {
  list-style: none;
  padding-left: 2.5ch;
  border-left: 1px solid var(--tw-prose-bullets);
  margin-left: 3ch;
  line-height: 1.875rem;
}

.markdown ul li a {
  font-family: var(--tw-prose-font-sans-serif);
  font-feature-settings: normal;
  font-weight: 500;
  font-size: 1rem;
  letter-spacing: var(--bdo-ls-tiny);
}

.markdown ul li a:first-child {
  font-weight: 550;
  font-size: 1.1875rem;
  letter-spacing: var(--bdo-ls-small);
}

.markdown ul li {
  color: var(--tw-prose-captions);
  font-size: 1rem;
  padding: 0.5rem 0;
}
</style>

<style>
/* Software TOC buttons */
.software-toc {
    border: 1px solid #e2e8f0;
    /* slate-200 */
    border-radius: 0.4rem;
    padding: 1rem;
    background-color: var(--tw-bg);
}

.software-toc .row {
    margin: -0.5rem;
    /* Compensate for inner padding */
}

.software-toc .col-md-6 {
    padding: 0.5rem;
}

.software-toc .btn-outline-nav {
    font-family: var(--tw-prose-font-sans-serif);
    font-feature-settings: normal;
    font-weight: 500;
    font-size: 1rem;
    letter-spacing: var(--bdo-ls-tiny);
    text-decoration: none;
    border: 1px solid #e2e8f0;
    /* slate-200 */
    border-radius: 0.4rem;
    padding: 1rem 1.25rem;
    height: 100%;
    text-align: left;
    background-color: #FFFFFF;
    display: flex;
    align-items: center;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.04), 0 1px 4px -1px rgba(0, 0, 0, 0.03);
    transition: background-color 0.15s ease-in-out, border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out, transform 0.15s ease-in-out;
}

.software-toc .btn-outline-nav:hover {
    background-color: #F9FAFB;
    /* Slightly off-white */
    border-color: #cbd5e1;
    /* Darker border, slate-300 */
    text-decoration: none;
    box-shadow: 0 2px 4px -1px rgba(0, 0, 0, 0.06), 0 4px 5px 0 rgba(0, 0, 0, 0.06);
    transform: translateY(-2px);
}

.software-toc .btn-outline-nav:active {
    background-color: #F3F4F6;
    border-color: #94a3b8;
    /* slate-400 */
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.05);
    transform: translateY(-1px);
}

/* Icon styles */
.software-toc .btn-outline-nav i {
    font-size: 1.1rem;
    width: 1.5rem;
    text-align: center;
    color: #475569;
    /* Primary-like color (slate-600) */
    margin-right: 0.75rem;
    transition: color 0.15s ease-in-out;
}

/* Text styles */
.software-toc .btn-outline-nav span {
    color: #475569;
    /* Readable dark gray (slate-600) */
    transition: color 0.15s ease-in-out;
}

.software-toc .btn-outline-nav:hover span {
    color: #1f2937;
    /* Darker text on hover (slate-800) */
}

.software-toc .btn-outline-nav:hover i {
    color: #1f2937;
    /* Darker primary on hover (slate-800) */
}

/* Adjust spacing on mobile */
@media (max-width: 767.98px) {
    .software-toc {
        padding: 0.75rem;
    }

    .software-toc .row {
        margin: -0.375rem;
    }

    .software-toc .col-md-6 {
        padding: 0.375rem;
    }
}
</style>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.7.2/css/all.min.css" integrity="sha512-Evv84Mr4kqVGRNSgIGL/F/aIDqQb7xQ2vcrdIwxfjThSH8CSR7PBEakCr51Ck+w+/U6swU2Im1vVX0SVk9ABhg==" crossorigin="anonymous" referrerpolicy="no-referrer" />
