---
title: "Software"
menu: "main"
---

<hr>

## R packages

I build R packages to improve my workflow.

**Infrastructure and automation**

- [pkglite](https://merck.github.io/pkglite/) <br> Compact package representations <br> [CRAN](https://cran.r-project.org/package=pkglite) | [GitHub](https://github.com/Merck/pkglite) | [Paper](https://doi.org/10.1177/17407745221123244)

**Statistical machine learning**

- [msaenet](https://nanx.me/msaenet/) <br> Multi-step adaptive estimation for sparse regressions <br> [CRAN](https://cran.r-project.org/package=msaenet) | [GitHub](https://github.com/nanxstats/msaenet) | [Paper](https://nanx.me/papers/msaenet.pdf)
- [stackgbm](https://nanx.me/stackgbm/) <br> Minimalist implementation of model stacking for gradient boosted tree models built by xgboost, lightgbm, and catboost <br> [GitHub](https://github.com/nanxstats/stackgbm)
- [oneclust](https://nanx.me/oneclust/) <br> Maximum homogeneity clustering for univariate data <br> [CRAN](https://cran.r-project.org/package=oneclust) | [GitHub](https://github.com/nanxstats/oneclust)
- [logreg](https://nanx.me/logreg/) <br> Regularized logistic regressions with computational graphs <br> [GitHub](https://github.com/nanxstats/logreg)
- [OHPL](https://OHPL.io/doc/) <br> Ordered homogeneity pursuit lasso for group variable selection <br> [CRAN](https://cran.r-project.org/package=OHPL) | [GitHub](https://github.com/nanxstats/OHPL) | [Paper](https://nanx.me/papers/OHPL.pdf)
- [RECA](https://nanx.me/RECA/) <br> Relevant component analysis for supervised distance metric learning <br> [CRAN](https://cran.r-project.org/package=RECA) | [GitHub](https://github.com/nanxstats/RECA)
- [enpls](https://nanx.me/enpls/) <br> Ensemble partial least squares regression <br> [CRAN](https://cran.r-project.org/package=enpls) | [GitHub](https://github.com/nanxstats/enpls)

**Statistical graphics and computing**

- [ggsci](https://nanx.me/ggsci/) <br> Scientific journal and sci-fi themed color palettes for ggplot2 <br> [CRAN](https://cran.r-project.org/package=ggsci) | [GitHub](https://github.com/nanxstats/ggsci)
- [liftr](https://nanx.me/liftr/) <br> Containerize R Markdown documents <br> [CRAN](https://cran.r-project.org/package=liftr) | [GitHub](https://github.com/nanxstats/liftr) | [Paper](https://doi.org/10.32614/RJ-2020-007)
- [hdnom](https://nanx.me/hdnom/) <br> Benchmarking and visualization toolkit for penalized Cox models <br> [CRAN](https://cran.r-project.org/package=hdnom) | [GitHub](https://github.com/nanxstats/hdnom) | [Paper](https://doi.org/10.1101/065524)

**Bioinformatics and cheminformatics**

- [protr](https://nanx.me/protr/) <br> Protein sequence feature extraction <br> [CRAN](https://cran.r-project.org/package=protr) | [GitHub](https://github.com/nanxstats/protr) | [Paper](https://nanx.me/papers/protr.pdf)
- [Rcpi](https://nanx.me/Rcpi/) <br> Feature extraction for proteins, compounds, and their interactions <br> [Bioconductor](https://bioconductor.org/packages/Rcpi) | [GitHub](https://github.com/nanxstats/Rcpi) | [Paper](https://nanx.me/papers/Rcpi.pdf)
- [tidychem](https://nanx.me/tidychem/) <br> Read, preprocess, and featurize chemical data using RDKit <br> [GitHub](https://github.com/nanxstats/tidychem)
- [grex](https://nanx.me/grex/) <br> Gene ID mapping for Genotype-Tissue Expression (GTEx) data <br> [CRAN](https://cran.r-project.org/package=grex) | [GitHub](https://github.com/nanxstats/grex)
- [cats](https://nanx.me/cats/) <br> Joint power analysis for non-symmetric two-stage case-control designs <br> [GitHub](https://github.com/nanxstats/cats)
- [ssw-r](https://nanx.me/ssw-r/) <br> R interface for SSW, a fast implementation of the Smith-Waterman algorithm using SIMD <br> [GitHub](https://github.com/nanxstats/ssw-r)
- [tidycwl](https://sbg.github.io/tidycwl/) <br> Tidy Common Workflow Language tools and workflows <br> [CRAN](https://cran.r-project.org/package=tidycwl) | [GitHub](https://github.com/sbg/tidycwl) | [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [biocompute](https://sbg.github.io/biocompute/) <br> Create and manipulate BioCompute Objects <br> [CRAN](https://cran.r-project.org/package=biocompute) | [GitHub](https://github.com/sbg/biocompute) | [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [sevenbridges-r](https://sbg.github.io/sevenbridges-r/) <br> Seven Bridges API client <br> [Bioconductor](https://bioconductor.org/packages/sevenbridges) | [GitHub](https://github.com/sbg/sevenbridges-r)
- [sbpack-r](https://nanx.me/sbpack-r/) <br> Pack and upload CWL apps to Seven Bridges powered platforms <br> [GitHub](https://github.com/nanxstats/sbpack-r)

## Shiny apps

Shiny apps, widgets, and templates for interactive data analysis.

- [shiny-loading-skeleton](https://github.com/nanxstats/shiny-loading-skeleton) <br> Shiny app template with an animated, fully customizable skeleton loader. <br> [Demo](https://nanx.shinyapps.io/shiny-loading-skeleton/)
- [shiny-fcp-loader](https://github.com/nanxstats/shiny-fcp-loader) <br> Shiny app template with a loading screen that aims to minimize the time to first contentful paint (FCP). <br> [Demo](https://nanx.shinyapps.io/shiny-fcp-loader/)
- [shiny-highlight-rmarkdown](https://github.com/nanxstats/shiny-highlight-rmarkdown) <br> Syntax highlighting for R Markdown output in Shiny based on highlight.js <br> [Demo](https://nanx.shinyapps.io/shiny-highlight-rmarkdown/)
- [exp2vec-shiny](https://github.com/nanxstats/exp2vec-shiny) <br> Shiny app for gene neighbors and gene analogies discovery with tissue-specific gene embeddings training on GTEx data. <br> [Demo](https://nanx.shinyapps.io/exp2vec/)
- [bco-app](https://github.com/sbg/bco-app) <br> Create, validate, and browse BioCompute Objects. Advanced track winner of the [precisionFDA challenge on BioCompute Objects](https://precision.fda.gov/challenges/7/results). <br> [Documentation](https://sbg.github.io/bco-app/) | [Paper](https://doi.org/10.12688/f1000research.25902.1)
- [protrweb](https://github.com/nanxstats/protrweb) <br> Compute protein sequence-derived descriptors <br> [Mirror 1](https://nanx.app/protr/) | [Mirror 2](https://nanx.shinyapps.io/protrweb/) | [Paper](https://nanx.me/papers/protr.pdf)
- [targetnet](https://github.com/nanxstats/targetnet) <br> Drug target identification with large-scale binding affinities data <br> [Mirror 1](https://nanx.app/targetnet/) | [Mirror 2](https://nanx.shinyapps.io/targetnet/)
- [imgsvd](https://github.com/nanxstats/imgsvd) <br> Image compression via singular value decomposition <br> [Mirror 1](https://nanx.app/imgsvd/) | [Mirror 2](https://nanx.shinyapps.io/imgsvd/)
- [hdnom-app](https://github.com/nanxstats/hdnom-app) <br> Benchmarking and visualization of penalized Cox models <br> [Mirror 1](https://nanx.app/hdnom/) | [Mirror 2](https://nanx.shinyapps.io/hdnom-app/)
- [conjugate-normal-umkv](https://github.com/nanxstats/conjugate-normal-umkv) <br> A demo of [Bayesian inference for normal mean (known variance)](https://stephens999.github.io/fiveMinuteStats/shiny_normal_example.html) <br> [Mirror 1](https://nanx.app/conjugate-normal-umkv/) | [Mirror 2](https://nanx.shinyapps.io/conjugate-normal-umkv/)
- [signify](https://github.com/nanxstats/signify) <br> Make your (>0.05) <em>p</em>-values "sound significant" <br> [Mirror 1](https://nanx.app/signify/) | [Mirror 2](https://nanx.shinyapps.io/signify/)
- [cwl-formatter](https://github.com/nanxstats/cwl-formatter) <br> Format Common Workflow Language (CWL) apps <br> [Mirror 1](https://nanx.app/cwl-formatter/)
- [cwl-exploder](https://github.com/nanxstats/cwl-exploder) <br> Unpack monolithic CWL workflows <br> [Mirror 1](https://nanx.app/cwl-exploder/)
- [cwl-upgrader](https://github.com/nanxstats/cwl-upgrader) <br> Upgrade CWL apps from sbg:draft2 to v1.0 <br> [Mirror 1](https://nanx.app/cwl-upgrader/)

## Data analysis pipelines

Reusable pipelines for analyzing real-world datasets in specific domains.

- [exp2vec](https://github.com/nanxstats/exp2vec) <br>
Pipeline for training tissue-specific gene embeddings on GTEx data.
- [bcpm-msaenet](https://github.com/nanxstats/bcpm-msaenet) <br>
Pipeline for the [precisionFDA challenge on brain cancer predictive modeling and biomarker discovery](https://precision.fda.gov/challenges/8/results). The 2nd place solution by predictive performance. Features multi-step adaptive SCAD-net and relaxed stability selection for feature selection, and a two-layer stacking tree model with xgboost, lightgbm, and catboost.
- [rankv](https://github.com/nanxstats/rankv) <br>
Pipeline for the [precisionFDA challenge on detecting adverse event anomalies using FDA open data](https://precision.fda.gov/challenges/9). Features a rank aggregated signal detection method for Vaccine Adverse Event Reporting System (VAERS) data, with base signal detection methods including Gamma Poisson Shrinker, PRR, ROR, and BCPNN. <br>
[Website](https://nanx.me/rankv/)

## Code contributor

- [gMCPLite](https://github.com/Merck/gMCPLite) <br> Lightweight fork of gMCP for graph-based multiple comparison procedures <br> [CRAN](https://cran.r-project.org/package=gMCPLite)
- [gsDesign Shiny app](https://rinpharma.shinyapps.io/gsdesign/) <br> Group sequential clinical trial design <br> [Mirror 1](https://gsdesign.shinyapps.io/prod/) | [Mirror 2](https://gsdesign.shinyapps.io/devel/)
- [PDXNet Portal](https://portal.pdxnetwork.org/) <br> Patient-derived xenograft (PDX) model, data, workflow, and tool discovery <br> [Paper](https://doi.org/10.1093/narcan/zcac014)
- [ashr](https://github.com/stephens999/ashr) <br> Methods for adaptive shrinkage, using Empirical Bayes <br> [CRAN](https://cran.r-project.org/package=ashr) | [Paper](https://doi.org/10.1093/biostatistics/kxw041)
- [dml](https://github.com/terrytangyuan/dml) <br> Distance metric learning in R <br> [CRAN](https://cran.r-project.org/package=dml) | [Paper](https://doi.org/10.21105/joss.01036)

## Code recipes & awesome lists

- [awesome-shiny-extensions](https://github.com/nanxstats/awesome-shiny-extensions) <br> A curated list of R packages that offer extended UI or server components for Shiny.
- [deep-learning-recipes](https://github.com/nanxstats/deep-learning-recipes) <br> R implementation for selected machine learning methods with deep learning frameworks.

## Visual Studio Code extensions

- [vscode-textmate-rstheme](https://marketplace.visualstudio.com/items?itemName=nanxstats.textmate-rstheme) <br> A light theme for Visual Studio Code inspired by the TextMate (default) theme in RStudio IDE. <br> [GitHub repository](https://github.com/nanxstats/vscode-textmate-rstheme)

## Hugo/blogdown themes

- [hugo-tanka](https://github.com/nanxstats/hugo-tanka) <br> A minimalist theme for Hugo and blogdown <br> [Live Preview](https://nanx.me/hugo-tanka/) | [Hugo Themes](https://themes.gohugo.io/hugo-tanka/) | [blogdown book](https://bookdown.org/yihui/blogdown/other-themes.html)
- [hugo-renga](https://github.com/nanxstats/hugo-renga) <br> A clean Bootstrap 4 theme for Hugo and blogdown <br> [Live Preview](https://nanx.me/hugo-renga/)
- [hugo-nanx2020](https://github.com/nanxstats/hugo-nanx2020) <br> An opinionated Bootstrap 4 theme for Hugo and blogdown
- [hugo-nanx2021](https://github.com/nanxstats/hugo-nanx2021) <br> A minimalist Bootstrap 5 theme for Hugo and blogdown
- [hugo-nanx2022](https://github.com/nanxstats/hugo-nanx2022) <br> A minimalist Bootstrap 5 theme for Hugo and blogdown

## Collections

- [dockflow.org](https://dockflow.org) <br> Bioconductor workflow containerization with liftr <br> [GitHub](https://github.com/nanxstats/dockflow)
- [nanx.app](https://nanx.app) <br> A collection of shiny apps built by me <br> [GitHub](https://github.com/nanxstats/nanx.app)

## Archives

- [使用 R 和 Apache 开发 Web 应用程序](https://nanx.me/rapache/) (2011)
- [来自 Google 的 R 语言编码风格指南](https://nanx.me/rstyle/) (2011)
