---
title: "protr 1.7-0"
author: "Nan Xiao"
date: "2023-11-10T21:00:00"
slug: protr-1-7-0
categories: []
tags:
  - R
  - R packages
  - protein sequence
  - feature engineering
meta_img: "image/google-deepmind-Fv39DqWqtHw-unsplash.jpg"
description: >
  protr 1.7-0 brings new tools for parallel protein sequence similarity
  computation between two sets of proteins based on sequence alignment.
---



<div class="float">
<img src="/image/google-deepmind-Fv39DqWqtHw-unsplash.jpg" alt="3D render from Google DeepMind. Artist: Wes Cockx." />
<div class="figcaption">3D render from <a href="https://unsplash.com/photos/a-model-of-a-building-made-out-of-legos-Fv39DqWqtHw">Google DeepMind</a>. Artist: Wes Cockx.</div>
</div>
<p>I am glad to announce the release of <a href="https://nanx.me/protr/">protr</a> 1.7-0.
Install it from CRAN with:</p>
<pre class="r"><code>install.packages(&quot;protr&quot;)</code></pre>
<p>You can also install it from
<a href="https://anaconda.org/conda-forge/r-protr">conda-forge</a> in Python.</p>
<p>First released in 2012, protr was my very first open source R package.
The package generates representations for protein sequences such as
numerical features and similarity measures.
Since its inception, protr has evolved after 21 CRAN releases
by incorporating user feedback. With 296 Google Scholar
<a href="https://scholar.google.com/scholar?cites=14211153676111619910">citations</a>
as of November 2023, it has demonstrated to be a useful
component for many workflows in drug discovery and machine learning research.</p>
<div id="whats-new" class="section level2">
<h2>What’s new</h2>
<p>A simple but common task in computational biology is to compute similarities
between proteins based on their sequences. The capability to compute such
similarities efficiently is essential for research in drug discovery,
protein-protein interaction prediction, and biomarker identification, providing
a foundation for kernel-based machine learning and deep learning methods.</p>
<p>protr 1.7-0 adds a new function
<a href="https://nanx.me/protr/reference/crossSetSim.html"><code>crossSetSim()</code></a>
for computing sequence-derived similarity (using multiple sequence alignment)
for all combinations of <span class="math inline">\(m \times n\)</span> sequences from two protein sets of
size <span class="math inline">\(m\)</span> and <span class="math inline">\(n\)</span>. Here is an example:</p>
<pre class="r"><code>library(protr)

# Find example UniProt FASTA files
fasta_files &lt;- list.files(
  system.file(&quot;protseq&quot;, package = &quot;protr&quot;),
  pattern = &quot;^[A-Z][0-9][A-Z0-9]{3}[0-9]\\.fasta$&quot;,
  full.names = TRUE
)

# Select 10 and 500 FASTA files for the two protein sets
set.seed(42)
protein_set_1 &lt;- sample(fasta_files, size = 10, replace = TRUE)
protein_set_2 &lt;- sample(fasta_files, size = 500, replace = TRUE)

# Read sequences from the FASTA files for both protein sets
seq_set_1 &lt;- sapply(protein_set_1, readFASTA, USE.NAMES = FALSE)
seq_set_2 &lt;- sapply(protein_set_2, readFASTA, USE.NAMES = FALSE)

# Compute the similarity between the two sets of protein sequences in parallel
seq_sim &lt;- crossSetSim(seq_set_1, seq_set_2, cores = 8)

# Set the column and row names of the similarity matrix to proteins names
colnames(seq_sim) &lt;- names(seq_set_1)
rownames(seq_sim) &lt;- names(seq_set_2)

# Preview the similarity matrix
head(seq_sim)[, 1:2]</code></pre>
<p>This function extends the existing functions for computing pairwise
similarities within a single set of proteins.
The API is made consistent with the previous implementations,
using foreach and doParallel for parallelization.</p>
<p>With the advancements in modern R parallel computing infrastructure,
particularly in futureverse, these functions should be ideally written in
a parallel backend agnostic style using <code>%dofuture%</code> so the parallel context
can be configured in user land code and any available backend types can be used.
See the articles linked in <a href="https://github.com/nanxstats/r-future-recipes">nanxstats/r-future-recipes</a>
for more details.</p>
</div>
<div id="relevant-tools-for-sequence-similarity-in-protr" class="section level2">
<h2>Relevant tools for sequence similarity in protr</h2>
<p>In protr, two functions for computing pairwise similarity within
a single list of protein sequences are the most frequently used:</p>
<ul>
<li><a href="https://nanx.me/protr/reference/parSeqSim.html"><code>parSeqSim()</code></a> -
Computes the <span class="math inline">\(n^2/2 - n\)</span> pairwise similarities within a single
protein sequence set of size <span class="math inline">\(n\)</span> in parallel.
Our <a href="https://doi.org/10.1002/psp4.12002">multiple evidence fusion paper</a>
published in 2015 leveraged this similarity information to generate
features inspired by collaborative filtering for a classification algorithm.</li>
<li><a href="https://nanx.me/protr/reference/parSeqSimDisk.html"><code>parSeqSimDisk()</code></a> -
Sometimes, the sequence data and similarity matrices are too large
to fit into RAM—especially when in parallel.
This is a memory-aware version that splits the task into batches,
writes the results in each batch to disk, and eventually merges the
cached results together.</li>
</ul>
</div>
<div id="acknowledgments" class="section level2">
<h2>Acknowledgments</h2>
<p>I want to thank <a href="https://github.com/seb-mueller">Sebastian Mueller</a>
for raising the need and contributing the code
(<a href="https://github.com/nanxstats/protr/issues/34">nanxstats/protr#34</a>).
Also, a big shout out to the user community and
package developers (3 reverse imports) for your continued trust and support.</p>
</div>
